### PR TITLE
feat(openai): add support for all endpoints

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Tips for reviewers

**note: 99% of the line changes in this PR are non-code, in the form of snapshot files, riot lockfiles, test data (audio/image files), and test cassette files used to mock request/responses with the OpenAI API, which store binary image/audio file data in `.yaml` files that Github counts as lines of code.**

The major files that reviewers should focus on are:
- `ddtrace/contrib/openai/patch.py` - outlines which endpoint hooks are added, minor changes to overall patching to minimize coupling with endpoint hook information.
- `ddtrace/contrib/openai/_endpoint_hooks.py` - implementation of each endpoint hook, as well as minor refactor to minimize code duplication between endpoint hooks
- `tests/contrib/openai/test_openai.py`. - shows how each endpoint is expected to be used.

## Summary
This PR adds tracing support for the remaining endpoints not implemented by #5488. This includes:
- Model (list, retrieve)
- Edits (create)
- Images (create, edit, create variation)
- Audio (transcribe, translate)
- Files (list, create, delete, retrieve, retrieve contents)
- Fine-tunes (list, create, retrieve, cancel, list events, delete fine tuned model)
- Moderations (create)

Changes in this PR include:
- Modifying the design of the endpoint hooks to minimize duplicate code, and adding new endpoint hooks for the above endpoints.
- Ensuring rate-limit related metrics are tagged as numeric metrics rather than string tags.
- Ensure user's OpenAI API key is set as a tag in multiple use scenarios (set in env var, set as attribute in `openai`, directly added as a request param)
- Adding and correcting names of attributes extracted from `openai` (e.g. `organization`, `api_type`)
- Simplifying and separating patch logic from endpoint-specific tracing logic (only operationID is propagated from endpoint hook level to patch level)
- Moving some global integration-level tags to request-level (e.g. `openai.request.model`, `openai.request.endpoint`, added `openai.request.method`)

### Overall Patch Implementation

**note**: the overall patch implementation has not changed; this is simply a refresher on how the overall integration works.

The openai integration features patching via generators - each of the openAI API's endpoint methods (ex: `openai.Completions.create()`) is wrapped by `_patched_endpoint(endpoint_hook, ...)`, which itself takes a corresponding endpoint hook as an argument. A rough description of what happens when `_patched_endpoint()` is called is as follows:
- `_patched_endpoint(endpoint_hook, ...)` starts a trace and starts a generator `_traced_endpoint(endpoint_hook, ...)`
- `_traced_endpoint(endpoint_hook, ...)` starts `endpoint_hook().handle_request(...)`, which performs endpoint-specific tracing logic, and yields back to `_patched_endpoint()` once the request has been processed by the endpoint hook.
- `_patched_endpoint()` runs the underlying openai API method, then yields back the response/error back to `_traced_endpoint()`, which in turn yields the response/error back to the `endpoint_hook`.
- `endpoint_hook()` performs the endpoint-specific tracing logic for the response/error, then finishes the trace.

### Endpoint Hook Design

The internal implementation design of the endpoint hooks have been modified (building off of the refactor from #5865) to minimize code duplication. This means that each endpoint hook now stores the following:
- `_request_arg_params` and `_request_kwarg_params`, which are tuples containing the arg/kwarg signature of the underlying openAI API endpoint method. These tuples are used in `_EndpointHook._record_request(...)` to add request arg/kwarg parameters as span tags/metrics.
- `ENDPOINT_NAME`, `REQUEST_TYPE`, `OPERATION_ID` constants, which reflect the base endpoint name (e.g. `/completions`), http request type (e.g. `POST`), and operation ID as specified by the openAI API specifcations (e.g. `createCompletion`). The operationID is used as the span resource name, and the endpoint/request_type values are added as span tags.
- Each endpoint hook also features (optional) a `_record_request(...)` and `_record_response(...)` to add any endpoint-specific span tagging logic.

## Testing

The testing for this PR involves snapshot testing for each endpoint, as well as `vcrpy` cassettes used to mock request/responses to OpenAI. 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] OPTIONAL: PR description includes explicit acknowledgement of the performance implications of the change as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: releasenotes/notes/feat-openai-extra-endpoints-000fe24c40b5544d.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    openai: Adds support for metrics, logs, and traces for the models, edits, images, audio, files, fine-tunes, and
+      moderations endpoints. See `the docs <https://ddtrace.readthedocs.io/en/stable/integrations.html#openai>`_
+      for more information.
```
Comment (mabdinur) [inline releasenotes/notes/feat-openai-extra-endpoints-000fe24c40b5544d.yaml:6]:
```suggestion
      for more information.

```

File: riotfile.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
         Venv(
             pys=["3"],
             pkgs={
-                "flake8": ">=3.8,<3.9",
+                "flake8": "==3.8.4",
```
Comment (mabdinur) [inline riotfile.py:129]: Is this change intentional?

File: setup.cfg [hunk 3]
```diff
@@ -2,7 +2,8 @@
 universal=1
 
 [codespell]
-skip = *.json,*.cpp,*.c,.riot,.tox,.mypy_cache,.git,*ddtrace/vendor
+skip = *.json,*.cpp,*.c,.riot,.tox,.mypy_cache,.git,*ddtrace/vendor,tests/contrib/openai/cassettes/*
+exclude-file = .codespellignorelines
```
Comment (mabdinur) [inline setup.cfg:6]: Do we need a codespell ignores?

**Review by mabdinur — COMMENTED** [review]
Left some comments but I will probably need to do another indepth pass. @Yun-Kim  Can we schedule time to go through this together? That can probably speed up the code review.

Also what's the test coverage for our openapi integration? We support many endpoints and set a wide-range of tags for only specific scenarios. I think we should aim for 100% test coverage here. I am sure we will miss things in code review, this can reduce the risk of merging this PR.

**Review by mabdinur — COMMENTED** [review]
Left some comments but I will probably need to do another indepth pass. @Yun-Kim  Can we schedule time to go through this together? That can probably speed up the code review.

Also what's the test coverage for our openapi integration? We support many endpoints and set a wide-range of tags for only specific scenarios. I think we should aim for 100% test coverage here. I am sure we will miss things in code review, this can reduce the risk of merging this PR.

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 4]
```diff
@@ -3,25 +3,68 @@
...
+        """Set base-level openai tags, as well as request params from args and kwargs."""
+        endpoint = self.ENDPOINT_NAME
+        if endpoint is None:
+            endpoint = "%s" % args[0].OBJECT_NAME
+        span.set_tag_str("openai.request.endpoint", "/%s/%s" % (API_VERSION, endpoint))
+        span.set_tag_str("openai.request.method", self.REQUEST_TYPE)
+
+        base_level_tag_args = ("api_base", "api_type", "api_version")
+
+        if self._request_arg_params and len(self._request_arg_params) > 1:
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py:29]: The problem with that approach is that there is so much inconsistency in the openai-python library in regards to what args/kwargs are common to different endpoints, as well as whether the same arg is a kwarg for different endpoints, and common methods for certain endpoints. Ideally I would go with your approach (and that was my first approach at this), but the inconsistency of openai-python would result in a *lot* of duplicate code.

I think for now we can stick with this, and if their library improves in consistency, then we can consider shifting the design like you mentioned.

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 5]
```diff
@@ -3,25 +3,68 @@
...
+            for idx, arg in enumerate(self._request_arg_params[1:], 1):
+                if idx >= len(args):
+                    break
+                if arg is None or args[idx] is None:
+                    continue
+                if arg in base_level_tag_args:
+                    span.set_tag("openai.%s" % arg, args[idx] or "")
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py:36]: This is intentional - this would only get tagged if the user specifically specified an empty value. The above line checks that the arg is not a None value.

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 6]
```diff
@@ -3,25 +3,68 @@
...
+class _EndpointHook:
+    # Assume base arg signature follows openai.EngineAPIResource.create(...)
+    _request_arg_params = ("cls", "api_key", "api_base", "api_type", "request_id", "api_version", "organization")
+    _request_kwarg_params = ()
+    _prompt_completion = False
+    ENDPOINT_NAME = "openai"
+    REQUEST_TYPE = ""
+    OPERATION_ID = ""
+
+    def _record_request(self, span, integration, args, kwargs):
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py:21]: We cover all endpoints with tests, so we can be confident of 100% coverage here.

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 7]
```diff
@@ -3,25 +3,68 @@
...
+                        span.set_tag_str("openai.request.%s" % arg, integration.trunc(args[idx].name))
+                    elif isinstance(args[idx], bytes):  # For binary file data
+                        span.set_tag_str("openai.request.%s" % arg, "")
+                    else:
+                        span.set_tag("openai.request.%s" % arg, integration.trunc(args[idx]))
+        for kw_attr in self._request_kwarg_params:
             if kw_attr in kwargs:
                 if isinstance(kwargs[kw_attr], dict):
                     set_flattened_tags(
                         span, [("openai.request.{}.{}".format(kw_attr, k), v) for k, v in kwargs[kw_attr].items()]
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py]: other than `api_key` which itself is obfuscated before being tagged, I don't see any sensitive data that we should flag.

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 8]
```diff
@@ -241,25 +297,490 @@ def _post_response(self, pin, integration, span, args, kwargs, resp, error):
...
+        """
+        Embedding endpoint allows multiple inputs, each of which we specify a request tag for, so have to
+        manually set them in _pre_response().
+        """
+        text_input = kwargs.get("input")
+        if isinstance(text_input, list) and not isinstance(text_input[0], int):
+            for idx, inp in enumerate(text_input):
+                span.set_tag_str("openai.request.input.%d" % idx, integration.trunc(str(inp)))
+        else:
+            span.set_tag("openai.request.input", text_input)
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py]: Good catch, thanks!

File: riotfile.py (hunk 2 above)
Comment (Yun-Kim) [reply riotfile.py]: Nope, good catch.

File: setup.cfg (hunk 3 above)
Comment (Yun-Kim) [reply setup.cfg:6]: Yes, one of the tests specifically includes a typo on purpose which codespell flags as a typo.

File: ddtrace/contrib/openai/_endpoint_hooks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py]: On second thought, we don't need this empty value default as this would be tagged if the value is not empty.

File: ddtrace/contrib/openai/_endpoint_hooks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py:44]: On second thought, I think this might increase maintainability if we follow your suggestion. I've changed the EndpointHook design such that all EndpointHooks need to call the super method, which will automatically tag all text/numeric/dict request params, and each endpoint hook will need to handle all special case params.

**Review by mabdinur — DISMISSED** [review]

**Review by mabdinur — APPROVED** [review]
🚢

**Review by Kyle-Verhoog — APPROVED** [review]
lgtm from the guild

**Review by Kyle-Verhoog — DISMISSED** [review]

File: ddtrace/contrib/openai/_endpoint_hooks.py [hunk 9]
```diff
@@ -1,34 +1,75 @@
...
+                    continue
+                if arg in self._base_level_tag_args:
+                    span.set_tag("openai.%s" % arg, args[idx])
+                elif arg == "organization":
+                    span.set_tag_str("openai.organization.id", args[idx])
+                elif arg == "api_key":
+                    span.set_tag_str("openai.user.api_key", _format_openai_api_key(args[idx]))
                 else:
-                    span.set_tag("openai.request.%s" % kw_attr, kwargs[kw_attr])
+                    span.set_tag("openai.request.%s" % arg, args[idx])
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/openai/_endpoint_hooks.py]: Sounds good. I was able to confirm through manual testing that setting `span.set_tag_str(key, str(value))` for numeric values still allows querying by numeric value on the UI, so this shouldn't change anything functionally.

**Review by mabdinur — APPROVED** [review]
